Android/Tapp/app/src/main/java/app/anish/com/tapp/dialogs/LinkedInWebViewLoginDialog.java

```diff
@@ -0,0 +1,294 @@
+                    String msg = "State token doesn't match. Possible CSRF Attach";
+                    dismissDialogInError(msg);
+                    return true;
+                }
+
+                // If user doesn't allow authorization to our application,
+                // the authorizationToken will be null
+                String authorizationToken = uri.getQueryParameter(RESPONSE_TYPE_VALUE);
+                if (authorizationToken == null) {
+                    dismissDialogInError("The user didn't allow authorization");
+                    return true;
+                }
+
+                Log.i(LOG_TAG, "Auth token received " + authorizationToken);
+                String accessTokenUrl = getAccessTokenUrl(authorizationToken);
+
+                // Make the request in an Async Task
+                new PostRequestAsyncTask().execute(accessTokenUrl);
+            } else {
+                // default behaviour
+                String url = uri.toString();
+                Log.i(LOG_TAG, "Webview redirecting to : " + uri);
+                mWebView.loadUrl(url);
+            }
+
+            return true;
+        }
+    }
+
+    private class PostRequestAsyncTask extends AsyncTask<String, Void, Throwable> {
+
+        private volatile AccessToken accessToken;
+
+        @Override
+        protected void onPreExecute() {
+            if (mProgressDialog == null || !mProgressDialog.isShowing()) {
+                mProgressDialog = ProgressDialog.show(getContext(), "", getContext().getString(R.string.loading), true);
+            }
+        }
+
+        @Override
+        protected void onPostExecute(Throwable throwable) {
+            if (mProgressDialog != null && mProgressDialog.isShowing()) {
+                mProgressDialog.dismiss();
+            }
+
+            if (throwable != null) {
+                dismissDialogInError(throwable.getMessage());
+            } else {
+                dismissDialogInSuccess(accessToken);
+            }
+        }
+
+        @Override
+        protected Throwable doInBackground(String... urls) {
+            if (urls.length > 0) {
+                String url = urls[0];
+                try {
+                    accessToken = getLinkedinToken(url);
+                    Log.d("LINKEDIN TOKEN", "LinkedIn Token is " + accessToken.toString());
+                    return null;
+                } catch (IOException e) {
+                    return new Exception("Error Http response " + e.getLocalizedMessage());
+                } catch (JSONException e) {
+                    return new Exception("Error Parsing Http response " + e.getLocalizedMessage());
+                }
+            }
+
+            return new Exception("No urls provided to task");
+        }
+
+        /**
+         * Extracts token from URL and saves it to Shared Preferences
+         * @param url
+         */
+        private AccessToken getLinkedinToken (String url) throws IOException, JSONException{
```

Comment:
Your life could be simplified here as well. Rather than directly working with JSONObject, you can use a JSON serialization library in concert with a type adapter to convert network responses transparently into Java objects. Here is what I recommend:
1. Gson: https://github.com/google/gson
2. auto-value-gson: https://github.com/rharter/auto-value-gson

Gson is a great library for doing JSON serialization/deserialization of java objects to/from JSON. It works great as a converter for retrofit. auto-value-gson is an annotation processor code generator that makes your network models and their serialization 100% declaration and 0% implementation.

It's pretty tough to just jump straight into using these libraries because its difficult to just be told what this does for you. If you want, I can help you get started with it.